log/sequencer.go

```diff
@@ -82,9 +84,47 @@ func (s Sequencer) sequenceLeaves(mt *merkle.CompactMerkleTree, leaves []trillia
 	return nodeMap, sequenceNumbers
 }
 
-// Can possibly improve by deferring a function that attempts to rollback, which will
-// fail if the tx was committed. Should only do this if we can hide the details of
-// the underlying storage transactions.
+func (s Sequencer) initMerkleTreeFromStorage(currentRoot trillian.SignedLogRoot, tx storage.LogTX) (*merkle.CompactMerkleTree, error) {
+	var merkleTree *merkle.CompactMerkleTree
+
+	if currentRoot.TreeSize == 0 {
+		// This should be an empty tree
+		if currentRoot.TreeRevision > 0 {
+			return nil, fmt.Errorf("MT has zero size but non zero revision: %v", *merkleTree)
+		}
+		return merkle.NewCompactMerkleTree(s.hasher), nil
+	} else {
+		// Initialize the compact tree state to match the latest root in the database
```

Comment:
Sorry, I keep forgetting that this is Go, not some kind of structured programming language.